xds/src/main/java/io/grpc/xds/XdsLbState.java

```diff
@@ -129,17 +151,279 @@ final XdsComms shutdownAndReleaseXdsComms() {
     return xdsComms;
   }
 
+  @VisibleForTesting // Interface of weighted Round-Robin algorithm that is convenient for test.
+  interface WrrAlgorithm {
+    @Nullable
+    Locality pickLocality(List<LocalityState> wrrList);
+  }
+
+  private static final class WrrAlgorithmImpl implements WrrAlgorithm {
```

Comment:
I was thinking using the existing `RoundRobinLoadBalancer` to do RR in a locality.

Drop can be added by wrapping the `RoundRobinLoadBalancer` inside a delegating "PercentageDroppingLoadBalancer". The delegating picker would do a pre-pick decision on whether to drop before calling into the underlying picker.

Inter-locality WRR can be formulated into a "WrrCompositingPicker", which takes a list of per-locality pickers.

The compositing approach can maximize code re-usability, and move a lot of logic outside of `XdsLbState` and make them individually testable.

Reply:
Done in #5550.